File: src/omnia/data/structure/immutable/ImmutableSortedSet.java
package omnia.data.structure.immutable;

import java.util.Comparator;
import java.util.Iterator;
import java.util.Optional;
import java.util.stream.Stream;
import omnia.data.structure.Collection;
import omnia.data.structure.SortedSet;
import omnia.data.structure.mutable.TreeSet;

/** An immutable version of {@link SortedSet}. */
public final class ImmutableSortedSet<E> implements SortedSet<E> {
  private final TreeSet<E> baseSet;

  private static final ImmutableSortedSet<?> EMPTY_SET =
      new ImmutableSortedSet<>((o1, o2) -> 0, ImmutableSet.empty());

  @SuppressWarnings("unchecked")
  public static <E> ImmutableSortedSet<E> empty() {
    return (ImmutableSortedSet<E>) EMPTY_SET;
  }

  public static <E> ImmutableSortedSet<E> copyOf(
      Comparator<? super E> comparator, Collection<? extends E> other) {
    return new ImmutableSortedSet<>(comparator, other);
  }

  private ImmutableSortedSet(Comparator<? super E> comparator, Collection<? extends E> other) {
    this.baseSet = TreeSet.create(comparator);
    this.baseSet.addAll(other);
  }

  @Override
  public boolean contains(E element) {
    return baseSet.contains(element);
  }

  /**
   * @inheritDoc
   *
   * <b>Note:</b> this type-unsafe version of {@link #contains} is inherently less performant
   * because we cannot perform a binary search on objects of unknown types.
   *
   * @param element the element to search for
   * @return true if the set contains the element, false if not
   */
  @Override
  public boolean containsUnknownTyped(Object element) {
    return baseSet.containsUnknownTyped(element);
  }

  @Override
  public int count() {
    return baseSet.count();
  }

  @Override
  public boolean isPopulated() {
    return baseSet.isPopulated();
  }

  @Override
  public Optional<E> itemPreceding(E other) {
    return baseSet.itemPreceding(other);
  }

  @Override
  public Optional<E> itemFollowing(E other) {
    return baseSet.itemFollowing(other);
  }

  @Override
  public Iterator<E> iterator() {
    return baseSet.iterator();
  }

  @Override
  public Stream<E> stream() {
    return baseSet.stream();
  }
}

File: src/omnia/data/structure/immutable/ImmutableSortedSet.kt
package omnia.data.structure.immutable

import omnia.data.structure.Collection
import omnia.data.structure.SortedSet
import omnia.data.structure.mutable.SortedArraySet

/** An immutable version of [SortedSet]. */
class ImmutableSortedSet<E : Any>
private constructor(comparator: Comparator<in E>, other: Collection<out E>) : SortedSet<E> {

  private val baseSet: SortedArraySet<E> = SortedArraySet.create(comparator)

  init {
    baseSet.addAll(other)
  }

  override fun contains(item: E): Boolean {
    return baseSet.contains(item)
  }

  override fun containsUnknownTyped(item: Any?): Boolean {
    return baseSet.containsUnknownTyped(item)
  }

  override fun count() = baseSet.count()

  override val isPopulated get() = baseSet.isPopulated

  override fun itemPreceding(other: E): E? {
    return baseSet.itemPreceding(other)
  }

  override fun itemPrecedingUnknownTyped(other: Any?): E? {
    return baseSet.itemPrecedingUnknownTyped(other)
  }

  override fun itemFollowing(other: E): E? {
    return baseSet.itemFollowing(other)
  }

  override fun itemFollowingUnknownTyped(other: Any?): E? {
    return baseSet.itemFollowingUnknownTyped(other)
  }

  override fun iterator(): Iterator<E> {
    return baseSet.iterator()
  }

  companion object {

    private val EMPTY_SET: ImmutableSortedSet<*> =
      ImmutableSortedSet({ _, _ -> 0 }, ImmutableSet.empty<Any>())

    @JvmStatic
    fun <E : Any> empty(): ImmutableSortedSet<E> {
      @Suppress("UNCHECKED_CAST")
      return EMPTY_SET as ImmutableSortedSet<E>
    }

    @JvmStatic
    fun <E : Any> copyOf(
      comparator: Comparator<in E>, other: Collection<out E>,
    ): ImmutableSortedSet<E> {
      return ImmutableSortedSet(comparator, other)
    }
  }

}